### PR TITLE
[dag] concurrent dag store

### PR DESCRIPTION
**Pull request by ibalajiarun**
### Description

- Separated in-memory component of Dag store from persistence.
- `PersistentDagStore` handles both persistence and maintains the in-memory DAG store.

### CODE REVIEW

**Comment by ibalajiarun** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11505?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `main`
  * **PR #11362** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11362?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #11407** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11407?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #11505** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11505?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
        * **PR #11433** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11433?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11505?utm_source=stack-comment).

File: consensus/src/dag/dag_store.rs [hunk 1]
```diff
@@ -408,3 +373,213 @@ impl Dag {
...
+    }
+
+    pub fn read(&self) -> DagReadGuard {
+        DagReadGuard {
+            gc_guard: self.gc_guard.read(),
+            inner: self.inner.read(),
+        }
+    }
+
+    pub fn write(&self) -> DagWriteGuard {
```
Comment (sasha8) [inline consensus/src/dag/dag_store.rs:471]: I think this is redundant

File: consensus/src/dag/dag_store.rs
Comment (sasha8) [inline consensus/src/dag/dag_store.rs:1]: nit:
You could take 
let round_ref = self
            .nodes_by_round
            .entry(round)
            .or_insert_with(|| vec![None; self.author_to_index.len()]);
            out of validate and lock only that.

**Review by sasha8 — APPROVED** [review]

**Comment by sasha8** [comment]
Much cleaner now!

File: consensus/src/dag/dag_store.rs [hunk 2]
```diff
@@ -133,14 +88,29 @@ impl Dag {
...
+        let node = Arc::new(node);
+        let round_ref = self
+            .get_node_ref_mut(node.round(), node.author())
+            .expect("must be present");
+        ensure!(round_ref.is_none(), "race during insertion");
+        *round_ref = Some(NodeStatus::Unordered(node.clone()));
+        Ok(())
+    }
+
+    fn validate_new_node(&mut self, node: &CertifiedNode) -> anyhow::Result<()> {
```
Comment (zekun000) [inline consensus/src/dag/dag_store.rs:107]: hmm, it looks a bit weird to have &mut self for validation, but probably it makes insertion simpler

File: consensus/src/dag/dag_store.rs [hunk 3]
```diff
@@ -408,3 +372,114 @@ impl Dag {
         self.nodes_by_round.is_empty() && self.start_round > 1
     }
 }
+
+pub struct PersistentDagStore {
```
Comment (zekun000) [inline consensus/src/dag/dag_store.rs:376]: I'd call this DagStore and Dag -> InMemDag

File: consensus/src/dag/dag_store.rs [hunk 4]
```diff
@@ -133,14 +88,29 @@ impl Dag {
...
+    pub fn add_node_for_test(&mut self, node: CertifiedNode) -> anyhow::Result<()> {
+        self.validate_new_node(&node)?;
+        self.add_validated_node(node)
+    }
+
+    fn add_validated_node(&mut self, node: CertifiedNode) -> anyhow::Result<()> {
+        let node = Arc::new(node);
+        let round_ref = self
+            .get_node_ref_mut(node.round(), node.author())
+            .expect("must be present");
```
Comment (zekun000) [inline consensus/src/dag/dag_store.rs:101]: hmm, this is not true if pruning happens in between?

Comment (ibalajiarun) [reply consensus/src/dag/dag_store.rs:101]: yes, found this recently as well. we need a ensure! here for the round here. Also, need to fix the lowest_round() and highest_round() implementations.

File: consensus/src/dag/dag_store.rs (hunk 2 above)
Comment (ibalajiarun) [reply consensus/src/dag/dag_store.rs:107]: yes, it makes insertion and also last validation step simpler.

**Review by zekun000 — APPROVED** [review]

**Pull request merged**